File: examples/cpp/freespace_move_example.cpp
// Robot Raconteur Tesseract Service Example: Freespace Move Example C++
//
// This is an example client program showing how to use the Tesseract
// service with Robot Raconteur. It expects the abb_irb2400 example
// scene from tesseract_support package to be loaded into the service.
//
// The service should be started prior to running this program. The service
// can be started using the docker image or from the command line.
// See the readme for instructions on how to start the service
// using the docker image. From the command line, use the following command.
// Adjust /ws/install/share/tesseract_support to the location of the tesseract_support.
//
//  export TESSERACT_RESOURCE_PATH=/ws/install/share
//  tesseract_robotraconteur_service --urdf-file=/ws/install/share/tesseract_support/urdf/abb_irb2400.urdf \
//     --srdf-file=/ws/install/share/tesseract_support/urdf/abb_irb2400.srdf \
//     --task-plugin-config-file=/ws/devel/share/tesseract_task_composer/config/task_composer_plugins.yaml
//
// Robot Raconteur with C++ requires the RobotRaconteurCore, RobotRaconteurCompanion, and the generated
// "thunk" source from the service definition "robdef" files that are specific to the Tesseract
// service. The CMake command ROBOTRACONTEUR_GENERATE_THUNK is used to generate
// the "thunk" source files automatically at compile time. Because C++ is statically typed,
// the service definitions must be available at compile time. This is unlike Python where
// the service definitions are loaded at runtime.
//
// See the Robot Raconteur documentation for more information on using Robot Raconteur with C++.

#include <iostream>
#include "robotraconteur_generated.h"
#include <boost/uuid/uuid_generators.hpp>
#include <RobotRaconteurCompanion/Converters/EigenConverters.h>

// Convenience namespace aliases
namespace RR=RobotRaconteur;

namespace rr_planning = experimental::tesseract_robotics::tasks::planning;
namespace rr_tasks = experimental::tesseract_robotics::tasks;
namespace RRC_Eigen=RobotRaconteur::Companion::Converters::Eigen;
namespace rr_tesseract = experimental::tesseract_robotics;
namespace rr_common = experimental::tesseract_robotics::common;
namespace rr_command = experimental::tesseract_robotics::command_language;


// TODO: Add this function to the Robot Raconteur companion library
com::robotraconteur::uuid::UUID new_random_uuid()
{
    boost::uuids::uuid uuid = boost::uuids::random_generator()();
    com::robotraconteur::uuid::UUID ret;
    memcpy(ret.s.uuid_bytes.data(), &uuid, sizeof(uuid));
    return ret;
}

// Helper function to build a move instruction from a waypoint
auto build_move_instr(const Eigen::Isometry3d& wp)
{
    // Create a Cartesian waypoint from the Eigen Isometry transform
    rr_command::CartesianWaypointPtr wp_ptr(new rr_command::CartesianWaypoint());
    // Use a companion helper function to convert the Eigen transform to a Robot Raconteur transform
    wp_ptr->transform = RRC_Eigen::ToTransform(wp);
    wp_ptr->lower_tolerance = RR::AllocateRRArray<double>(0);
    wp_ptr->upper_tolerance = RR::AllocateRRArray<double>(0);

    // Create a move instruction from the waypoint
    rr_command::MoveInstructionPtr move_instr(new rr_command::MoveInstruction());
    move_instr->waypoint = wp_ptr;
    move_instr->profile = "DEFAULT";
    move_instr->move_type = rr_command::MoveInstructionType::freespace;
    move_instr->uuid = new_random_uuid();
    return move_instr;
}

// Main entry point
int main(int argc, char **argv) {

    // Initialize Robot Raconteur using ClientNodeSetup, the generated thunk source types, and command line arguments
    RR::ClientNodeSetup node_setup(ROBOTRACONTEUR_SERVICE_TYPES, argc, argv);

    // Connect to the service. In this case, a localhost url is used. If the service were running
    // on a remote node, the URL would be "rr+tcp://hostname.example.com:port" for TCP. An IP
    // address can also be used directly.
    //
    // Robot Raconteur has advanced discovery and subscriptions that can be used instead of URLs. See
    // the documentation for more information.
    rr_tesseract::TesseractRoboticsPtr c = RR::rr_cast<rr_tesseract::TesseractRobotics>(RR::RobotRaconteurNode::s()->ConnectService("rr+tcp://localhost:63158?service=tesseract"));

    // Get the available task pipeline information from the service. Tesseract task pipelines
    // execute planning and other operations
    auto task_info = c->get_task_pipelines_info();
    std::string output_key = RR::RRArrayToString(task_info->at("FreespacePipeline")->output_keys->front());

    // Create the environment. In this case, the environment is loaded from the service. The
    // available environments are specified when the service is started. "default" is the
    // default environment.
    c->load_environment("default", "env");

    // Create the instructions
    auto instr1 = build_move_instr(Eigen::Isometry3d::Identity() * Eigen::Translation3d(0.9, -0.3, 1.455) * Eigen::Quaterniond(0.70710678,0,0.70710678,0));
    auto instr2 = build_move_instr(Eigen::Isometry3d::Identity() * Eigen::Translation3d(0.9, 0.3, 1.455) * Eigen::Quaterniond(0.70710678,0,0.70710678,0));

    // Create and configure the manipulator info structure
    rr_common::ManipulatorInfoPtr manip_info(new rr_common::ManipulatorInfo());
    manip_info->manipulator = "manipulator";
    manip_info->tcp_frame = "tool0";
    manip_info->working_frame = "base_link";

    // Create and configure the composite instruction
    rr_command::CompositeInstructionPtr composite(new rr_command::CompositeInstruction());
    composite->order = rr_command::CompositeInstructionOrder::ordered;
    composite->instructions = RR::AllocateEmptyRRList<RR::RRValue>();
    composite->instructions->push_back(instr1);
    composite->instructions->push_back(instr2);
    composite->profile = "DEFAULT";
    composite->manipulator_info = manip_info;
    composite->uuid = new_random_uuid();

    // Create and configure the planning problem
    rr_planning::PlanningTaskComposerProblemPtr problem(new rr_planning::PlanningTaskComposerProblem());
    problem->name = "example";
    problem->input = composite;
    problem->environment_name = "env";
    problem->manip_info = manip_info;

    // Get the default task executor. In this case, the task executor will be using Taskflow
    // to execute the pipeline
    auto executor = c->get_task_executors("default");

    // Create the task executor input and set the pipeline name. In this case, we are using
    // the FreespacePipeline. Other pipelines are available. These pipelines are configured
    // when the service is started.
    rr_tasks::TaskExecutorInputPtr input(new rr_tasks::TaskExecutorInput());
    input->problem = problem;
    input->pipeline_name = "FreespacePipeline";

    // Run the pipeline
    auto exec_gen = executor->run(input);

    // The pipeline returns a Robot Raconteur "Generator". Generators are a type of coroutine
    // that are used either for a long running operation or a sequence operation.
    // Repeat TryNext() until the generator is complete.
    rr_tasks::TaskExecutorStatusPtr status;
    bool res;
    do
    {
        res = exec_gen->TryNext(status);
    }
    while (res);

    // Check the status of the pipeline take make sure it completed successfully
    if (status->action_status != com::robotraconteur::action::ActionStatusCode::complete)
    {
        std::cerr << "Planning did not complete successfully" << std::endl;
        return 1;
    }

    // Retrieve the output plan from the data storage
    auto output = RR::rr_cast<rr_command::CompositeInstruction>(status->data_storage->at(output_key));

    // Extract the joint trajectory from the output plan
    std::vector<Eigen::VectorXd> joint_trajectory;
    for (auto instr : *output->instructions)
    {
        auto move_instr = RR::rr_cast<rr_command::MoveInstruction>(instr);
        auto wp = RR::rr_cast<rr_command::StateWaypoint>(move_instr->waypoint);
        joint_trajectory.push_back(RRC_Eigen::RRArrayToEigen<Eigen::VectorXd>(wp->position));
    }

    // Print out the joint trajectory
    for (auto& j : joint_trajectory)
    {
        std::cout <<"[" << j.transpose() << "]" << std::endl;
    }

    // TODO: plot result in viewer!

    return 0;
}

File: examples/python/freespace_move_example.py
# Robot Raconteur Tesseract Service Example: Freespace Move Example
#
# This is an example client program showing how to use the Tesseract
# service with Robot Raconteur. It expects the abb_irb2400 example
# scene from tesseract_support package to be loaded into the service.
#
# The service should be started prior to running this program. The service
# can be started using the docker image or from the command line.
# See the readme for instructions on how to start the service
# using the docker image. From the command line, use the following command.
# Adjust /ws/install/share/tesseract_support to the location of the tesseract_support.
#
#  export TESSERACT_RESOURCE_PATH=/ws/install/share
#  tesseract_robotraconteur_service --urdf-file=/ws/install/share/tesseract_support/urdf/abb_irb2400.urdf \
#     --srdf-file=/ws/install/share/tesseract_support/urdf/abb_irb2400.srdf \
#     --task-plugin-config-file=${workspaceFolder}/ws/devel/share/tesseract_task_composer/config/task_composer_plugins.yaml
#
# Once the service is running, the client can be run from another terminal. The robotraconteur and 
# robotraconteurcomanion python packages must be installed. Install using the following commands:
#
#    python3 -m pip install robotraconteur robotraconteurcompanion
#
# The client only requires the robotraconteur and robotraconteurcompanion python packages. The tesseract
# package is not required on the client side.
#
# To run the client use the following command:
#
#    python3 freespace_move_example.py
#
# Note the use of RR.VarValue throughout the example. This is required to specify the type of the
# variant data. This is required because Python does not carry enough information to automatically
# determine the type of the variant data.

# Import Robot Raconteur and general Python modules. The companion package is not
# required but contains some utility classes for working with Robot Raconteur that
# simplify working with advanced types.
from RobotRaconteur.Client import *
from RobotRaconteurCompanion.Util.GeometryUtil import GeometryUtil
from RobotRaconteurCompanion.Util.UuidUtil import UuidUtil
import numpy as np

def main():
    
    # Connect to the service. In this case, a localhost url is used. If the service were running
    # on a remote node, the URL would be "rr+tcp://hostname.example.com:port" for TCP. An IP
    # address can also be used directly.
    #
    # Robot Raconteur has advanced discovery and subscriptions that can be used instead of URLs. See
    # the documentation for more information.
    c = RRN.ConnectService('rr+tcp://localhost:63158?service=tesseract')

    # Initialize the helper classes from the companion.
    geom = GeometryUtil(client_obj=c)
    uuid_util = UuidUtil(client_obj=c)


    # Retrieve the constants and structure types from the service. Robot Raconteur provides
    # plug-and-play capabilities, so this information is provided at runtime by the service.
    # The client does not need to be configured with this information, it is provided
    # automatically by the service.
    command_lang_const = RRN.GetConstants("experimental.tesseract_robotics.command_language",c)
    action_const = RRN.GetConstants("com.robotraconteur.action",c)
    cart_wp_type = RRN.GetStructureType("experimental.tesseract_robotics.command_language.CartesianWaypoint",c)
    move_instr_type = RRN.GetStructureType("experimental.tesseract_robotics.command_language.MoveInstruction",c)
    composite_instr_type = RRN.GetStructureType("experimental.tesseract_robotics.command_language.CompositeInstruction",c)
    manip_info_type = RRN.GetStructureType("experimental.tesseract_robotics.common.ManipulatorInfo",c)
    planning_problem_type = RRN.GetStructureType("experimental.tesseract_robotics.tasks.planning.PlanningTaskComposerProblem",c)
    task_exec_input_type = RRN.GetStructureType("experimental.tesseract_robotics.tasks.TaskExecutorInput",c)

    # Get the available task pipeline information from the service. Tesseract task pipelines
    # execute planning and other operations
    task_info = c.task_pipelines_info
    output_key = task_info["FreespacePipeline"].output_keys[0]

    # Create the environment. In this case, the environment is loaded from the service. The
    # available environments are specified when the service is started. "default" is the
    # default environment.
    c.load_environment("default", "env")

    # Helper function to make a Move instruction
    def build_move_instr(xyz, rpy):
        # Create a cartesian waypoint. Joint waypoints are also available.
        wp = cart_wp_type()
        wp.transform = geom.xyz_rpy_to_transform(xyz,rpy)

        # Create a move instruction and set the waypoint
        instr = move_instr_type()
        instr.waypoint = RR.VarValue(wp, "experimental.tesseract_robotics.command_language.CartesianWaypoint")
        instr.profile = "DEFAULT"
        instr.move_type = command_lang_const["MoveInstructionType"]["freespace"]
        instr.uuid = uuid_util.NewRandomUuid()
        return RR.VarValue(instr, "experimental.tesseract_robotics.command_language.MoveInstruction")

    # Create the instructions
    instr1 = build_move_instr([0.8, -0.3, 1.455], np.deg2rad([0,90,0]))
    instr2 = build_move_instr([0.8, 0.3, 1.455], np.deg2rad([0,90,0]))
    instr3 = build_move_instr([0.8, 0.5, 1.455], np.deg2rad([0,90,0]))

    # Create and configure the manipulator info structure
    manip_info = manip_info_type()
    manip_info.tcp_frame = "tool0"
    manip_info.working_frame = "base_link"
    manip_info.manipulator = "manipulator"

    # Create and configure the composite instruction
    composite = composite_instr_type()
    composite.order = command_lang_const["CompositeInstructionOrder"]["ordered"]
    composite.instructions = [instr1, instr2] #, instr3]
    composite.manipulator_info = manip_info
    composite.uuid = uuid_util.NewRandomUuid()

    # Create and configure the planning problem
    problem = planning_problem_type()
    problem.name = "example"
    problem.input = RR.VarValue(composite, "experimental.tesseract_robotics.command_language.CompositeInstruction")
    problem.environment_name = "env"
    problem.manip_info = manip_info

    # Get the default task executor. In this case, the task executor will be using Taskflow
    # to execute the pipeline
    executor = c.get_task_executors("default")

    # Create the task executor input and set the pipeline name. In this case, we are using
    # the FreespacePipeline. Other pipelines are available. These pipelines are configured
    # when the service is started.
    exec_input = task_exec_input_type()
    exec_input.pipeline_name = "FreespacePipeline"
    exec_input.problem = RR.VarValue(problem, "experimental.tesseract_robotics.tasks.planning.PlanningTaskComposerProblem")

    # Run the pipeline
    exec_gen = executor.run(exec_input)

    # The pipeline returns a Robot Raconteur "Generator". Generators are a type of coroutine
    # that are used either for a long running operation or a sequence operation.
    # Repeat TryNext() until the generator is complete.
    status = None
    while True:
        res, status1 = exec_gen.TryNext()
        if not res:
            break
        status = status1
    
    # Check the status of the pipeline take make sure it completed successfully
    assert status.action_status == action_const["ActionStatusCode"]["complete"]

    # Retrieve the output plan from the data storage
    output_plan = status.data_storage[output_key]

    # Extract and print the planned waypoints
    wp_positions = []

    for instr_v in output_plan.data.instructions:
        instr = instr_v.data
        wp = instr.waypoint.data
        wp_positions.append(wp.position)

    print(wp_positions)

    # TODO: plot result in viewer!
        






if __name__ == '__main__':
    main()